transformers/storage/vat/repository_test.go

```diff
@@ -756,17 +770,69 @@ var _ = Describe("Vat storage repository", func() {
 
 				It("deletes urn state associated with diff if it's the earliest state in the table", func() {
 					initialArt := rand.Int()
-					setupErrOne := repo.Create(diffID, headerOne.Id, urnArtMetadata, strconv.Itoa(initialArt))
+					setupErrOne := repo.Create(diffIdOne, headerOne.Id, urnArtMetadata, strconv.Itoa(initialArt))
 					Expect(setupErrOne).NotTo(HaveOccurred())
 
-					_, deleteErr := db.Exec(deleteRowQuery, headerOne.Id)
+					_, deleteErr := db.Exec(deleteHeaderQuery, headerOne.Id)
 					Expect(deleteErr).NotTo(HaveOccurred())
 
 					var urnStates []test_helpers.UrnState
 					queryErr := db.Select(&urnStates, getArtQuery)
 					Expect(queryErr).NotTo(HaveOccurred())
 					Expect(len(urnStates)).To(Equal(0))
 				})
+
+				It("handles removing snapshots when there's a three-part reorg", func() {
+					// this test case mimics the issue that is detailed in VDB-1756 where we saw a reorg that resulted in
+					// receiving three headers - canonical, noncanonical, canonical
+
+					// set up the header, diff and snapshot before the scenario being tested
+					artBefore := rand.Int()
+					blockBefore := rand.Int()
+					blockBeforeTimestamp := int64(rand.Int31())
+
+					hashBefore := common.BytesToHash([]byte{0, 2, 4, 6, 8})
+					headerBefore := CreateHeaderWithHash(hashBefore.String(), blockBeforeTimestamp, blockBefore, db)
+					diffId0 := CreateFakeDiffRecordWithHeader(db, headerBefore)
+					beforeSetupErr := repo.Create(diffId0, headerBefore.Id, urnArtMetadata, strconv.Itoa(artBefore))
+					Expect(beforeSetupErr).NotTo(HaveOccurred())
+
+					// TEST SCENARIO: header with hashOne inserted and then removed in lieu of header with hashTwo
+					// then header with hashOne is reinserted as canonical and header with hashTwo is removed
+					art := artBefore + 1
+					block := blockBefore + 1
+					blockTimestamp := blockBeforeTimestamp + 1
+
+					// HEADER WITH HASH ONE
+					headerWithHashOne := CreateHeaderWithHash(hashOne.String(), blockTimestamp, block, db)
+					diffId1 := CreateFakeDiffRecordWithHeader(db, headerWithHashOne)
+					setupErrOne := repo.Create(diffId1, headerWithHashOne.Id, urnArtMetadata, strconv.Itoa(art))
+					Expect(setupErrOne).NotTo(HaveOccurred())
+
```

Comment:
Is it accurate that we'd expect the number of urn states to be 2 here as well, before the next header/diff for the same block is inserted?

Reply:
Yep, that's right. I'll add an assertion here too, just to be explicit.